Accept 429 response codes as an indication of the secondary rate limit being exceeded

holly-cummins opened the pull request:
Resolves #1842.
Resolves #1805. 

# Description

If a user uses this library to drive the GitHub API hard, there will quite often be failures caused by secondary rate limits being breached (#1842 and #1805). There’s been some discussion of headers on #1842, and also of pluggable 'handle the rate limit' modules, but I think there’s a more fundamental problem with the handling of rate limits. (The good news is this means there's a simpler solution, too.)

The `AbuseLimitHandler` checks for a rate limit violation by looking for a 403 header and some headers:  

```
@Override
boolean isError(@Nonnull GitHubConnectorResponse connectorResponse) {
    return isForbidden(connectorResponse) && hasRetryOrLimitHeader(connectorResponse);
}
```

`HTTP_FORBIDDEN` is a 403. The [GitHub docs](https://docs.github.com/en/rest/using-the-rest-api/rate-limits-for-the-rest-api?apiVersion=2022-11-28#exceeding-the-rate-limit) indicate that limits from GitHub are signalled with either 403 or `TOO_MANY_REQUESTS` responses (429). In my experience, I've only seen 429s.

### What about headers? 

The GitHub docs docs suggest that any headers are optional, and that in the absence of headers the code should just wait a minute (which, luckily, is what the current implementation already does):

>  If the retry-after response header is present, you should not retry your request until after that many seconds has elapsed. If the x-ratelimit-remaining header is 0, you should not retry your request until after the time, in UTC epoch seconds, specified by the x-ratelimit-reset header. Otherwise, wait for at least one minute before retrying. If your request continues to fail due to a secondary rate limit, wait for an exponentially increasing amount of time between retries, and throw an error after a specific number of retries.

Based on these docs, I think it's sufficient to count a 429 as a rate limit response, even if headers are missing. Semantically, a 429 is an indication of an abuse detection/rate limit scenario, and should be handled as such.

### What about the rate limit handler?

`GitHubRateLimitHandler` also requires a 403 to 'count' a response as rate limited:

```
@Override
boolean isError(@NotNull GitHubConnectorResponse connectorResponse) throws IOException {
connectorResponse.statusCode() == HttpURLConnection.HTTP_FORBIDDEN
            && "0".equals(connectorResponse.header("X-RateLimit-Remaining"));
```

I didn't update this one, since the secondary rate limit checks are now correctly being handled on the abuse path with my changes. It might be worth updating it, or also even considering whether the distinction between the abuse and rate limit paths is so fuzzy that having both is just maintenance overhead. But that’s a different discussion. 

# Before submitting a PR:

- [ X] Changes must not break binary backwards compatibility. If you are unclear on how to make the change you think is needed while maintaining backward compatibility, [CONTRIBUTING.md](CONTRIBUTING.md) for details.
- [ X] Add JavaDocs and other comments explaining the behavior. 
- [ ] When adding or updating methods that fetch entities, add `@link` JavaDoc entries to the relevant documentation on https://docs.github.com/en/rest . 
- [ ] Add tests that cover any added or changed code. This generally requires capturing snapshot test data. See [CONTRIBUTING.md](CONTRIBUTING.md) for details.
- [ ] Run `mvn -D enable-ci clean install site` locally. If this command doesn't succeed, your change will not pass CI.
- [ X] Push your changes to a branch other than `main`. You will create your PR from that branch.

# When creating a PR: 

- [X] Fill in the "Description" above with clear summary of the changes. This includes:
  - [X ] If this PR fixes one or more issues, include "Fixes #<issue number>" lines for each issue. 
  - [ X] Provide links to relevant documentation on https://docs.github.com/en/rest where possible. If not including links, explain why not.
- [X ] All lines of new code should be covered by tests as reported by code coverage. Any lines that are not covered must have PR comments explaining why they cannot be covered. For example, "Reaching this particular exception is hard and is not a particular common scenario."
- [X] Enable "Allow edits from maintainers".

bitwiseman commented on src/main/java/org/kohsuke/github/GitHubAbuseLimitHandler.java:
```diff
@@ -29,7 +29,19 @@ public abstract class GitHubAbuseLimitHandler extends GitHubConnectorResponseErr
+    }
+
+    /**
+     * Checks if the response status code is TOO_MANY_REQUESTS (403).
```
```suggestion
     * Checks if the response status code is TOO_MANY_REQUESTS (429).
```

bitwiseman commented on src/main/java/org/kohsuke/github/AbuseLimitHandler.java:
```diff
@@ -89,7 +89,7 @@ public void onError(IOException e, HttpURLConnection uc) throws IOException {
             String v = uc.getHeaderField("Retry-After");
             if (v == null)
-                return 60 * 1000; // can't tell, return 1 min
+                return 61 * 1000; // can't tell, return 1 min plus 1 second for luck
```
Is there a functional reason for this?

holly-cummins replied:
The [GitHub docs](https://docs.github.com/en/rest/using-the-rest-api/rate-limits-for-the-rest-api?apiVersion=2022-11-28#exceeding-the-rate-limit) say "Otherwise, wait for at least one minute before retrying." I never know if the check is `< 60s` or `<=60s`, but it seems like waiting something that is unambiguously more than a minute is safest.

bitwiseman reviewed (changes requested):
This is a definite improvement.  

It still needs tests. Look at the existing abuse limit tests, which use manually constructed test data. Maybe just add one more step to the existing test rather than adding a whole new test. Up to you.

holly-cummins commented:
Oh good, I'm glad it looks sensible. I've added some tests and squashed.

bitwiseman commented on src/main/java/org/kohsuke/github/AbuseLimitHandler.java:
```diff
@@ -89,7 +89,7 @@ public void onError(IOException e, HttpURLConnection uc) throws IOException {
             String v = uc.getHeaderField("Retry-After");
             if (v == null)
-                return 60 * 1000; // can't tell, return 1 min
+                return 61 * 1000; // can't tell, return 1 min plus 1 second for luck
```
```suggestion
                // can't tell, wait for unambiguously over one minute per GitHub guidance
                return 61 * 1000;
```

bitwiseman commented:
Please don't squash/force-push.  We can squash when merging the PR.

bitwiseman reviewed (changes requested):
Test failure needs fixing.

holly-cummins commented:
> Test failure needs fixing.

Oops, sorry about that. And noted about the squashing - different orgs have different requirements, so I have a squash habit, but I'll keep the commits separate. I agree it does make subsequent reviews easier!

bitwiseman commented on src/test/java/org/kohsuke/github/AbuseLimitHandlerTest.java:
```diff
@@ -412,7 +412,6 @@ public void onError(IOException e, HttpURLConnection uc) throws IOException {
                         assertThat(uc.getContentType(), equalTo("application/json; charset=utf-8"));
                         assertThat(uc.getContentLength(), equalTo(-1));
                         assertThat(uc.getHeaderFields(), instanceOf(Map.class));
-                        assertThat(uc.getHeaderFields().size(), Matchers.greaterThan(25));
```
I wonder why I even had this in there...

holly-cummins replied:
Yeah, I think there's an opportunity to make some of the tests a bit more concise and flexible without losing the important sense checks about the mocks' behaviour or the validation of the actual code ... but I didn't want to change too much existing code in an unrelated PR. But I perhaps should have been a bit more conservative about how much I copy-pasted.

holly-cummins commented:
The CI's failing because I didn't have enough coverage, even with the new test. To get the `parseWaitTime` method to be testable, I had to do a bit of refactoring to make it visible to the test code. I also added a new test to cover both paths, and then I realised that the original code could fail with a `NumberFormatException`, because `Retry-After` can [be a date](https://www.rfc-editor.org/rfc/rfc9110#field.retry-after). So I added new code and tests to handle that case. I also did a bit of refactoring of the test code and removing assertions about the mock that I was pretty sure we definitely didn't care about.

bitwiseman reviewed (approved):
Sorry for the slow response cycle.

bitwiseman commented on src/main/java/org/kohsuke/github/AbuseLimitHandler.java:
```diff
@@ -105,4 +101,26 @@ public void onError(IOException e, HttpURLConnection uc) throws IOException {
         }
     };
+
+    /*
+     * Exposed for testability. Given an http response, find the retry-after header field
+     * and parse it as either a number or a date (the spec allows both).
+     * If no header is found, wait for a reasonably amount of time.
+     */
```
```suggestion
    /*
     * Exposed for testability. Given an http response, find the retry-after header field and parse it as either a
     * number or a date (the spec allows both). If no header is found, wait for a reasonably amount of time.
     */
```

bitwiseman commented on src/test/java/org/kohsuke/github/AbuseLimitHandlerTest.java:
```diff
@@ -349,16 +342,191 @@ public void onError(IOException e, HttpURLConnection uc) throws IOException {
+        assertThat(mockGitHub.getRequestCount(), equalTo(2));
+    }
+
+    @Test
```
```suggestion
    /**
     * Tests the behavior of the GitHub API client when the abuse limit handler with a date retry.
     *
     * @throws Exception
     *             if any error occurs during the test execution.
     */
    @Test
```

bitwiseman commented on src/test/java/org/kohsuke/github/AbuseLimitHandlerTest.java:
```diff
@@ -349,16 +342,191 @@ public void onError(IOException e, HttpURLConnection uc) throws IOException {
+        assertThat(mockGitHub.getRequestCount(), equalTo(2));
+    }
+
+    @Test
```
```suggestion
    /**
     * Tests the behavior of the GitHub API client when the abuse limit handler with a no retry after header.
     *
     * @throws Exception
     *             if any error occurs during the test execution.
     */
    @Test
```

holly-cummins commented on src/main/java/org/kohsuke/github/AbuseLimitHandler.java:
```diff
@@ -105,4 +101,22 @@ public void onError(IOException e, HttpURLConnection uc) throws IOException {
+        } catch (NumberFormatException nfe) {
+            // The retry-after header could be a number in seconds, or an http-date
+            ZonedDateTime zdt = ZonedDateTime.parse(v, DateTimeFormatter.RFC_1123_DATE_TIME);
+            return ChronoUnit.MILLIS.between(ZonedDateTime.now(), zdt);
```
Thanks! (And thanks for explaining about the diff to the response time, that makes a lot of sense.)

bitwiseman replied:
Sure thing.  If you have time to circle back and address #1909 , that would be awesome.  Thanks for the great contribution!
